### DISCUSSION TITLE
Exposing registered components via entry points.

### DISCUSSION THREAD

**Discussion by coltonflowers1** [explosion/spaCy#12482] [Help: Coding & Implementations] [answered]
I would like to include a custom-registered function within a config.cfg without having to have users pass in the code for these registered functions to the training CLI via the `--code` argument. Entry points have allowed me to do this with component factories, and according to the official documentation [here](https://spacy.io/usage/saving-loading#entry-points), I can also do this with custom languages, displacy colors, and custom lookups, but there does not seem to be anything for plain registered functions. In fact, the docs [here](https://spacy.io/usage/processing-pipelines#component-data-initialization) only reference the `--code` solution in order to make sure your custom registered function are imported prior to training.

**Comment by adrianeboyd** [comment, marked as answer]
There is not a separate registry for components because components are just a wrapper for factory registry entries.

However the entry points behavior is pretty confusing because:

* importing code related to an entry point can run additional code through decorators
* there is a hidden internal entry points registry for factories just to make it easy to import+run code with decorators, and it's the decorators that are important not the entry point names

So with this code + entry point, you'll end up with a factory for the name `my_component`:

```python
from spacy import Language


@Language.component("my_component")
def my_method(doc):
    return doc
```

```ini
spacy_factories =
    _some_unused_name = my_package.components:my_method
```

The component would also have been registered in the same way if you'd added this as an entry point in `misc` or some other category, but it's less confusing if you're not adding extraneous entries to other registries.

This is pretty confusing (I was initially not 100% sure without testing this out), we should add some better examples to the docs...

**Reply by coltonflowers1** [reply]
My apologies, I edited the second link in my post as it was pointing to the wrong section of the docs. If I understand you correctly, you are saying that the `spacy_factories` end point can be used to store functions I define using the `misc` registry? I'm not actually trying to register a pipeline component but a function I use within the initialization of a custom component because I want more control over that function in my configuration. I assumed `spacy_factories` was only for pipeline components because of the following in the [docs](https://spacy.io/usage/saving-loading#entry-points): "[The spacy factories end point is a] group of entry points for pipeline component factories, keyed by component name. Can be used to expose custom components defined by another package."

**Reply by adrianeboyd** [reply]
I thought you mean a component with `Language.component` vs `Language.factory`. The registries in this link are the main ones with some custom internal behavior related to each registry: https://spacy.io/usage/saving-loading#entry-points

Most of the other registries are just to make the names easier to filter/remember, and the specific registry and entry point names don't actually matter as long as it returns a function with the right signature: https://spacy.io/api/top-level#registry

You could have:

```ini
scorer = {"@batchers": "my_corpus_reader.v1"}
```

I mean, I don't recommend doing this, but just to demonstrate.

Using `misc` is fine with:

```ini
spacy_misc =
    my_component.v1 = my_package.components:my_component
```

If you define the entry point in your package setup, then you don't need the decorator in the code.

As an example with a number of different entry points:

https://github.com/explosion/spacy-legacy/blob/eecb096ed01ead79a7f8d0fc64631becf37d4458/setup.cfg#L35-L57

**Reply by coltonflowers1** [reply]
I see so I am ok to refer to a function registered in any registry using entry points and the entry point name doesn't actually matter. I think that clears up my confusion. Thanks!

**Reply by adrianeboyd** [reply]
Just to clarify, the entry point name does matter for everything except `spacy_factories`. The registered function will show up with that name in the registry.
